themes/default/content/blog/pulumi-is-imperative-declarative-imperative/index.md

```diff
@@ -0,0 +1,205 @@
+
+Sketch the setup of the article as the description of the PR for reviewers
+-->
+
+On a regular basis, articles and tweets pass by discussing whether or not some specific tool is imperative or declarative.
+It's no surprise that Pulumi is often the tool being debated. What if I tell you that Pulumi is imperative, declarative and imperative?
+
+<!--more-->
+
+When we look at our [Frequently Asked Questions](/docs/support/faq/#is-pulumi-imperative-or-declarative), we read the following on the declarative versus imperative topic:
+
+> Pulumi is a declarative tool that uses imperative languages to define your end state. The language is used for authoring your program.
+> It’s not used for talking to the cloud provider API.
+
+The title of the article mentions `imperative` and `declarative` both. Pulumi leverages the best of both worlds
+into our product.
+
+{{< tweet id="1553431913691430912" user="adamhjk" >}}
+
+Pulumi tries to offer a solution where our customers are only limited by their imagination rather than the tool at hand.
+Let me use the Pulumi architecture to highlight why I mention *imperative* twice.
+
+## Pulumi Architecture
+
+If you don't know the Pulumi architecture, here is the diagram from our [How Pulumi Works](/docs/intro/concepts/how-pulumi-works/) page:
+
+![Pulumi Architecture](/images/docs/reference/engine-block-diagram.png)
+
+You code your infrastructure in your preferred programming language. When you are done coding, you run `pulumi up` and the Pulumi CLI starts
+the language host for your selected programming language, as well as the required providers. The interaction between these 3 parts of
+the architecture results in the actual creation or modification of your infrastructure.
+
+### Language Host
+
+Under the hood, our Pulumi CLI does a lot of things, but one of the first actions is starting the language runtime which is configured in
+the `Pulumi.yaml` project file. Here is a small Python example:
+
+```python
+from pulumi_aws import s3
+
+bucket = s3.Bucket('bucket')
+
+for i in range(10):
+    s3.BucketObject(
+        f'object-{i}',
+        s3.BucketObjectArgs(
+            bucket=bucket.id,
+            key=str(i),
+        )
+    )
+```
+
+A Pulumi program models the to-be state of your infrastructure. If you read the program above, you can see that
+we define 11 resources as our to be infrastructure:
+
+* 1 AWS S3 bucket
+* 10 Objects in the bucket created in the previous step
+
+While this is definitely an imperative program, there is one important thing to understand: instantiating an
+`s3.Bucket`, `s3.BucketObject` or any other Pulumi resource should not be interpreted as an imperative creation
+of the resource in the language host. Behind the scenes, any resource instantiation in the language host triggers a
+`Register Resource` request to the Pulumi engine.
+
+No matter what the actual state of your infrastructure is, running your program will always send to intended
+resource model to the engine.
```

Comment:
What is sent to the engine?  
Maybe re-order this sentence, it sounds confusing since it seems like we left out a word. 

`Running your program always sends the resource model to the Pulumi Engine regardless of what state your infrastructure is in` - you can reword it however you want.  I

Reply:
Done.